File: src/main/python/net/coderextreme/data/ViewFrustumPrototype.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.1"
head1 = x3d.head()
meta2 = x3d.meta()
meta2.name = "title"
meta2.content = "ViewFrustumPrototype.x3d"

head1.children.append(meta2)
meta3 = x3d.meta()
meta3.name = "description"
meta3.content = "Define a view frustum associated with a given pair of Viewpoint and NavigationInfo nodes, provided as a reusable prototype."

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "creator"
meta4.content = "Don Brutzman"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "translated"
meta5.content = "16 August 2008"

head1.children.append(meta5)
meta6 = x3d.meta()
meta6.name = "modified"
meta6.content = "Mon, 09 Feb 2026 07:13:15 GMT"

head1.children.append(meta6)
meta7 = x3d.meta()
meta7.name = "reference"
meta7.content = "ViewFrustumExample.x3d"

head1.children.append(meta7)
meta8 = x3d.meta()
meta8.name = "drawing"
meta8.content = "ViewFrustumComputation.png"

head1.children.append(meta8)
meta9 = x3d.meta()
meta9.name = "Image"
meta9.content = "ViewFrustumOverheadView.png"

head1.children.append(meta9)
meta10 = x3d.meta()
meta10.name = "Image"
meta10.content = "ViewFrustumObliqueView.png"

head1.children.append(meta10)
meta11 = x3d.meta()
meta11.name = "subject"
meta11.content = "view culling frustum"

head1.children.append(meta11)
meta12 = x3d.meta()
meta12.name = "identifier"
meta12.content = "https://X3dGraphics.com/examples/X3dForWebAuthors/Chapter14Prototypes/ViewFrustumPrototype.x3d"

head1.children.append(meta12)

X3D0.head = head1
Scene13 = x3d.Scene()
ProtoDeclare14 = x3d.ProtoDeclare()
ProtoDeclare14.name = "ViewFrustum"
ProtoDeclare14.appinfo = "Display view frustum associated with a given pair of Viewpoint NavigationInfo nodes"
ProtoInterface15 = x3d.ProtoInterface()
field16 = x3d.field()
field16.accessType = "initializeOnly"
field16.type = "SFNode"
field16.name = "ViewpointNode"
field16.appinfo = "required: insert Viewpoint DEF or USE node for view of interest"

ProtoInterface15.field.append(field16)
field17 = x3d.field()
field17.accessType = "initializeOnly"
field17.type = "SFNode"
field17.name = "NavigationInfoNode"
field17.appinfo = "required: insert NavigationInfo DEF or USE node of interest"

ProtoInterface15.field.append(field17)
field18 = x3d.field()
field18.accessType = "inputOutput"
field18.type = "SFBool"
field18.name = "visible"
field18.value = True
field18.appinfo = "whether or not frustum geometry is rendered"

ProtoInterface15.field.append(field18)
field19 = x3d.field()
field19.accessType = "inputOutput"
field19.type = "SFColor"
field19.name = "lineColor"
field19.value = [0.9,0.9,0.9]
field19.appinfo = "RGB color of ViewFrustum outline, default value 0.9 0.9 0.9"

ProtoInterface15.field.append(field19)
field20 = x3d.field()
field20.accessType = "inputOutput"
field20.type = "SFColor"
field20.name = "frustumColor"
field20.value = [0.8,0.8,0.8]
field20.appinfo = "RGB color of ViewFrustum hull geometry, default value 0.8 0.8 0.8"

ProtoInterface15.field.append(field20)
field21 = x3d.field()
field21.accessType = "inputOutput"
field21.type = "SFFloat"
field21.name = "transparency"
field21.value = 0.5
field21.appinfo = "transparency of ViewFrustum hull geometry, default value 0.5"

ProtoInterface15.field.append(field21)
field22 = x3d.field()
field22.accessType = "inputOutput"
field22.type = "SFFloat"
field22.name = "aspectRatio"
field22.value = 0.75
field22.appinfo = "assumed ratio height/width, default value 0.75"

ProtoInterface15.field.append(field22)
field23 = x3d.field()
field23.accessType = "initializeOnly"
field23.type = "SFBool"
field23.name = "trace"
field23.appinfo = "debug support, default false"

ProtoInterface15.field.append(field23)

ProtoDeclare14.ProtoInterface = ProtoInterface15
ProtoBody24 = x3d.ProtoBody()
Switch25 = x3d.Switch(DEF="VisibilitySwitch")
Transform26 = x3d.Transform(DEF="PositionTransform")
Transform26.rotation = [0,1,0,3.14159]
Transform27 = x3d.Transform(DEF="OrientationTransform")
Shape28 = x3d.Shape()
Appearance29 = x3d.Appearance()
Material30 = x3d.Material()
IS31 = x3d.IS()
connect32 = x3d.connect()
connect32.nodeField = "emissiveColor"
connect32.protoField = "lineColor"

IS31.connect.append(connect32)

Material30.IS = IS31

Appearance29.material = Material30

Shape28.appearance = Appearance29
IndexedLineSet33 = x3d.IndexedLineSet(DEF="FrustumLines")
IndexedLineSet33.coordIndex = [0,1,2,3,0,-1,4,5,6,7,4,-1,0,4,-1,1,5,-1,2,6,-1,3,7,-1]
Coordinate34 = x3d.Coordinate(DEF="FrustumCoordinate")
Coordinate34.point = [(0, 0, 0),(0, 0, 0),(0, 0, 0),(0, 0, 0),(0, 0, 0),(0, 0, 0),(0, 0, 0),(0, 0, 0)]

IndexedLineSet33.coord = Coordinate34

Shape28.geometry = IndexedLineSet33

Transform27.children.append(Shape28)
Shape35 = x3d.Shape()
Appearance36 = x3d.Appearance(DEF="FrustumAppearance")
Material37 = x3d.Material()
IS38 = x3d.IS()
connect39 = x3d.connect()
connect39.nodeField = "diffuseColor"
connect39.protoField = "frustumColor"

IS38.connect.append(connect39)
connect40 = x3d.connect()
connect40.nodeField = "transparency"
connect40.protoField = "transparency"

IS38.connect.append(connect40)

Material37.IS = IS38

Appearance36.material = Material37

Shape35.appearance = Appearance36
Extrusion41 = x3d.Extrusion(DEF="FrustumExtrusion")

Shape35.geometry = Extrusion41

Transform27.children.append(Shape35)
Shape42 = x3d.Shape()
Appearance43 = x3d.Appearance(USE="FrustumAppearance")

Shape42.appearance = Appearance43
Sphere44 = x3d.Sphere()
Sphere44.radius = 0.08

Shape42.geometry = Sphere44

Transform27.children.append(Shape42)

Transform26.children.append(Transform27)

Switch25.children.append(Transform26)

ProtoBody24.children.append(Switch25)
Script45 = x3d.Script(DEF="GeometryComputationScript")
Script45.url = ["ViewFrustumPrototypeScript.js"]
Script45.directOutput = True
field46 = x3d.field()
field46.accessType = "inputOutput"
field46.type = "SFBool"
field46.name = "visible"
field46.appinfo = "Whether or not frustum geometry is rendered"

Script45.field.append(field46)
field47 = x3d.field()
field47.accessType = "outputOnly"
field47.type = "SFInt32"
field47.name = "visibilitySwitchSelection"
field47.appinfo = "Adjust Switch selection to make geometry visible or not"

Script45.field.append(field47)
field48 = x3d.field()
field48.accessType = "initializeOnly"
field48.type = "SFNode"
field48.name = "ViewpointNode"

Script45.field.append(field48)
field49 = x3d.field()
field49.accessType = "initializeOnly"
field49.type = "SFNode"
field49.name = "NavigationInfoNode"

Script45.field.append(field49)
field50 = x3d.field()
field50.accessType = "initializeOnly"
field50.type = "SFNode"
field50.name = "FrustumCoordinate"
Coordinate51 = x3d.Coordinate(USE="FrustumCoordinate")

field50.children.append(Coordinate51)

Script45.field.append(field50)
field52 = x3d.field()
field52.accessType = "initializeOnly"
field52.type = "SFNode"
field52.name = "FrustumExtrusion"
Extrusion53 = x3d.Extrusion(USE="FrustumExtrusion")

field52.children.append(Extrusion53)

Script45.field.append(field52)
field54 = x3d.field()
field54.accessType = "inputOnly"
field54.type = "SFBool"
field54.name = "recompute"

Script45.field.append(field54)
field55 = x3d.field()
field55.accessType = "inputOutput"
field55.type = "SFFloat"
field55.name = "aspectRatio"
field55.appinfo = "assumed ratio height/width"

Script45.field.append(field55)
field56 = x3d.field()
field56.accessType = "outputOnly"
field56.type = "SFVec3f"
field56.name = "position_changed"

Script45.field.append(field56)
field57 = x3d.field()
field57.accessType = "outputOnly"
field57.type = "SFRotation"
field57.name = "orientation_changed"

Script45.field.append(field57)
field58 = x3d.field()
field58.accessType = "outputOnly"
field58.type = "MFVec3f"
field58.name = "spine_changed"

Script45.field.append(field58)
field59 = x3d.field()
field59.accessType = "outputOnly"
field59.type = "MFVec2f"
field59.name = "scale_changed"

Script45.field.append(field59)
field60 = x3d.field()
field60.accessType = "outputOnly"
field60.type = "MFVec3f"
field60.name = "point_changed"

Script45.field.append(field60)
field61 = x3d.field()
field61.accessType = "initializeOnly"
field61.type = "SFBool"
field61.name = "trace"

Script45.field.append(field61)
IS62 = x3d.IS()
connect63 = x3d.connect()
connect63.nodeField = "visible"
connect63.protoField = "visible"

IS62.connect.append(connect63)
connect64 = x3d.connect()
connect64.nodeField = "ViewpointNode"
connect64.protoField = "ViewpointNode"

IS62.connect.append(connect64)
connect65 = x3d.connect()
connect65.nodeField = "NavigationInfoNode"
connect65.protoField = "NavigationInfoNode"

IS62.connect.append(connect65)
connect66 = x3d.connect()
connect66.nodeField = "aspectRatio"
connect66.protoField = "aspectRatio"

IS62.connect.append(connect66)
connect67 = x3d.connect()
connect67.nodeField = "trace"
connect67.protoField = "trace"

IS62.connect.append(connect67)

Script45.IS = IS62

ProtoBody24.children.append(Script45)
ROUTE68 = x3d.ROUTE()
ROUTE68.fromNode = "GeometryComputationScript"
ROUTE68.fromField = "visibilitySwitchSelection"
ROUTE68.toNode = "VisibilitySwitch"
ROUTE68.toField = "set_whichChoice"

ProtoBody24.children.append(ROUTE68)
ROUTE69 = x3d.ROUTE()
ROUTE69.fromNode = "GeometryComputationScript"
ROUTE69.fromField = "position_changed"
ROUTE69.toNode = "PositionTransform"
ROUTE69.toField = "set_translation"

ProtoBody24.children.append(ROUTE69)
ROUTE70 = x3d.ROUTE()
ROUTE70.fromNode = "GeometryComputationScript"
ROUTE70.fromField = "orientation_changed"
ROUTE70.toNode = "OrientationTransform"
ROUTE70.toField = "set_rotation"

ProtoBody24.children.append(ROUTE70)
ROUTE71 = x3d.ROUTE()
ROUTE71.fromNode = "GeometryComputationScript"
ROUTE71.fromField = "spine_changed"
ROUTE71.toNode = "FrustumExtrusion"
ROUTE71.toField = "set_spine"

ProtoBody24.children.append(ROUTE71)
ROUTE72 = x3d.ROUTE()
ROUTE72.fromNode = "GeometryComputationScript"
ROUTE72.fromField = "scale_changed"
ROUTE72.toNode = "FrustumExtrusion"
ROUTE72.toField = "set_scale"

ProtoBody24.children.append(ROUTE72)
ROUTE73 = x3d.ROUTE()
ROUTE73.fromNode = "GeometryComputationScript"
ROUTE73.fromField = "point_changed"
ROUTE73.toNode = "FrustumCoordinate"
ROUTE73.toField = "set_point"

ProtoBody24.children.append(ROUTE73)

ProtoDeclare14.ProtoBody = ProtoBody24

Scene13.children.append(ProtoDeclare14)
WorldInfo74 = x3d.WorldInfo()
WorldInfo74.title = "ViewFrustumPrototype.x3d"

Scene13.children.append(WorldInfo74)
Anchor75 = x3d.Anchor()
Anchor75.description = "ViewFrustum Example"
Anchor75.url = ["ViewFrustumExample.x3d"]
Shape76 = x3d.Shape()
Appearance77 = x3d.Appearance()
Material78 = x3d.Material()
Material78.diffuseColor = [0.8,0.4,0]

Appearance77.material = Material78

Shape76.appearance = Appearance77
Text79 = x3d.Text()
Text79.string = ["ViewFrustumPrototype.x3d","is a Prototype declaration file.","For an example scene using the prototype,","click this text and view","ViewFrustumExample.x3d"]
FontStyle80 = x3d.FontStyle()
FontStyle80.size = 0.8
FontStyle80.justify = ["MIDDLE","MIDDLE"]

Text79.fontStyle = FontStyle80

Shape76.geometry = Text79

Anchor75.children.append(Shape76)

Scene13.children.append(Anchor75)

X3D0.Scene = Scene13
f = open("../data/ViewFrustumPrototype.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/ViewFrustumPrototype.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/ViewFrustumPrototype.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/node/net/ecma/data/ViewFrustumPrototype.mjs
'use strict';
var X3D = require('./x3d.mjs');
try {
	var fs = require('fs');
} catch (e) {
console.log("Problems loading fs. On browser?",e);
}
var SFString = require('./x3d.mjs');
var SFNode = require('./x3d.mjs');
var head = require('./x3d.mjs');
var MFNode = require('./x3d.mjs');
var meta = require('./x3d.mjs');
var Scene = require('./x3d.mjs');
var WorldInfo = require('./x3d.mjs');
var ProtoDeclare = require('./x3d.mjs');
var ProtoInterface = require('./x3d.mjs');
var field = require('./x3d.mjs');
var ProtoBody = require('./x3d.mjs');
var Switch = require('./x3d.mjs');
var SFInt32 = require('./x3d.mjs');
var Transform = require('./x3d.mjs');
var SFRotation = require('./x3d.mjs');
var Shape = require('./x3d.mjs');
var IndexedLineSet = require('./x3d.mjs');
var MFInt32 = require('./x3d.mjs');
var Coordinate = require('./x3d.mjs');
var MFVec3f = require('./x3d.mjs');
var Appearance = require('./x3d.mjs');
var Material = require('./x3d.mjs');
var IS = require('./x3d.mjs');
var connect = require('./x3d.mjs');
var Extrusion = require('./x3d.mjs');
var Sphere = require('./x3d.mjs');
var SFFloat = require('./x3d.mjs');
var Script = require('./x3d.mjs');
var SFBool = require('./x3d.mjs');
var MFString = require('./x3d.mjs');
var ROUTE = require('./x3d.mjs');
var Anchor = require('./x3d.mjs');
var SFColor = require('./x3d.mjs');
var Text = require('./x3d.mjs');
var FontStyle = require('./x3d.mjs');
var X3D0 =  new X3D({

      profile : new SFString("Immersive"),
      version : new SFString("4.0"),
      head : new SFNode(
        new head({
          meta : new MFNode([
            new meta({
              name : new SFString("title"),
              content : new SFString("ViewFrustumPrototype.x3d")}),

            new meta({
              name : new SFString("description"),
              content : new SFString("Define a view frustum associated with a given pair of Viewpoint and NavigationInfo nodes, provided as a reusable prototype.")}),

            new meta({
              name : new SFString("creator"),
              content : new SFString("Don Brutzman")}),

            new meta({
              name : new SFString("translated"),
              content : new SFString("16 August 2008")}),

            new meta({
              name : new SFString("modified"),
              content : new SFString("20 October 2019")}),

            new meta({
              name : new SFString("reference"),
              content : new SFString("ViewFrustumExample.x3d")}),

            new meta({
              name : new SFString("drawing"),
              content : new SFString("ViewFrustumComputation.png")}),

            new meta({
              name : new SFString("Image"),
              content : new SFString("ViewFrustumOverheadView.png")}),

            new meta({
              name : new SFString("Image"),
              content : new SFString("ViewFrustumObliqueView.png")}),

            new meta({
              name : new SFString("subject"),
              content : new SFString("view culling frustum")}),

            new meta({
              name : new SFString("identifier"),
              content : new SFString("https://X3dGraphics.com/examples/X3dForWebAuthors/Chapter14Prototypes/ViewFrustumPrototype.x3d")}),

            new meta({
              name : new SFString("generator"),
              content : new SFString("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit")}),

            new meta({
              name : new SFString("license"),
              content : new SFString("../license.html")})])})),
      Scene : new SFNode(
        new Scene({
          children : new MFNode([
            new WorldInfo({
              title : new SFString("ViewFrustumPrototype.x3d")}),

            new ProtoDeclare({
              name : new SFString("ViewFrustum"),
              appinfo : new SFString("Display view frustum associated with a given pair of Viewpoint NavigationInfo nodes"),
              ProtoInterface : new SFNode(
                new ProtoInterface({
                  field : new MFNode([
                    new field({
                      type : field.TYPE_SFNODE,
                      name : new SFString("ViewpointNode"),
                      accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                      appinfo : new SFString("required: insert Viewpoint DEF or USE node for view of interest"),
                      /*NULL node, ProtoInstance must provide*/}),

                    new field({
                      type : field.TYPE_SFNODE,
                      name : new SFString("NavigationInfoNode"),
                      accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                      appinfo : new SFString("required: insert NavigationInfo DEF or USE node of interest"),
                      /*NULL node, ProtoInstance must provide*/}),

                    new field({
                      type : field.TYPE_SFBOOL,
                      name : new SFString("visible"),
                      accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                      appinfo : new SFString("whether or not frustum geometry is rendered"),
                      value : new SFString("true")}),

                    new field({
                      type : field.TYPE_SFCOLOR,
                      name : new SFString("lineColor"),
                      accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                      appinfo : new SFString("RGB color of ViewFrustum outline, default value 0.9 0.9 0.9"),
                      value : new SFString("0.9 0.9 0.9")}),

                    new field({
                      type : field.TYPE_SFCOLOR,
                      name : new SFString("frustumColor"),
                      accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                      appinfo : new SFString("RGB color of ViewFrustum hull geometry, default value 0.8 0.8 0.8"),
                      value : new SFString("0.8 0.8 0.8")}),

                    new field({
                      type : field.TYPE_SFFLOAT,
                      name : new SFString("transparency"),
                      accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                      appinfo : new SFString("transparency of ViewFrustum hull geometry, default value 0.5"),
                      value : new SFString("0.5")}),

                    new field({
                      type : field.TYPE_SFFLOAT,
                      name : new SFString("aspectRatio"),
                      accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                      appinfo : new SFString("assumed ratio height/width, default value 0.75"),
                      value : new SFString("0.75")}),

                    new field({
                      type : field.TYPE_SFBOOL,
                      name : new SFString("trace"),
                      accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                      appinfo : new SFString("debug support, default false"),
                      value : new SFString("false")})])})),
              ProtoBody : new SFNode(
                new ProtoBody({
                  children : new MFNode([
                    new Switch({
                      DEF : new SFString("VisibilitySwitch"),
                      whichChoice : new SFInt32(-1),
                      children : new MFNode([
                        new Transform({
                          DEF : new SFString("PositionTransform"),
                          rotation : new SFRotation([0,1,0,3.14159]),
                          children : new MFNode([
                            new Transform({
                              DEF : new SFString("OrientationTransform"),
                              child : new SFNode(
                                new Shape({
                                  geometry : new SFNode(
                                    new IndexedLineSet({
                                      DEF : new SFString("FrustumLines"),
                                      coordIndex : new MFInt32([0,1,2,3,0,-1,4,5,6,7,4,-1,0,4,-1,1,5,-1,2,6,-1,3,7,-1]),
                                      coord : new SFNode(
                                        new Coordinate({
                                          DEF : new SFString("FrustumCoordinate"),
                                          point : new MFVec3f([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])}))})),
                                  appearance : new SFNode(
                                    new Appearance({
                                      material : new SFNode(
                                        new Material({
                                          IS : new SFNode(
                                            new IS({
                                              connect : new MFNode([
                                                new connect({
                                                  nodeField : new SFString("emissiveColor"),
                                                  protoField : new SFString("lineColor")})])}))}))}))})),
                              child : new SFNode(
                                new Shape({
                                  geometry : new SFNode(
                                    new Extrusion({
                                      DEF : new SFString("FrustumExtrusion")})),
                                  appearance : new SFNode(
                                    new Appearance({
                                      DEF : new SFString("FrustumAppearance"),
                                      material : new SFNode(
                                        new Material({
                                          IS : new SFNode(
                                            new IS({
                                              connect : new MFNode([
                                                new connect({
                                                  nodeField : new SFString("diffuseColor"),
                                                  protoField : new SFString("frustumColor")}),

                                                new connect({
                                                  nodeField : new SFString("transparency"),
                                                  protoField : new SFString("transparency")})])}))}))}))})),
                              child : new SFNode(
                                new Shape({
                                  geometry : new SFNode(
                                    new Sphere({
                                      radius : new SFFloat(0.08)})),
                                  appearance : new SFNode(
                                    new Appearance({
                                      USE : new SFString("FrustumAppearance")}))}))})])})])}),

                    new Script({
                      DEF : new SFString("GeometryComputationScript"),
                      directOutput : new SFBool(true),
                      url : new MFString(["ViewFrustumPrototypeScript.js"]),
                      field : new MFNode([
                        new field({
                          type : field.TYPE_SFBOOL,
                          name : new SFString("visible"),
                          accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                          appinfo : new SFString("Whether or not frustum geometry is rendered")}),

                        new field({
                          type : field.TYPE_SFINT32,
                          name : new SFString("visibilitySwitchSelection"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY),
                          appinfo : new SFString("Adjust Switch selection to make geometry visible or not")}),

                        new field({
                          type : field.TYPE_SFNODE,
                          name : new SFString("ViewpointNode"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                          /*initialization node (if any) goes here*/}),

                        new field({
                          type : field.TYPE_SFNODE,
                          name : new SFString("NavigationInfoNode"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                          /*initialization node (if any) goes here*/}),

                        new field({
                          type : field.TYPE_SFNODE,
                          name : new SFString("FrustumCoordinate"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                          children : new MFNode([
                            new Coordinate({
                              USE : new SFString("FrustumCoordinate")})])}),

                        new field({
                          type : field.TYPE_SFNODE,
                          name : new SFString("FrustumExtrusion"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                          children : new MFNode([
                            new Extrusion({
                              USE : new SFString("FrustumExtrusion")})])}),

                        new field({
                          type : field.TYPE_SFBOOL,
                          name : new SFString("recompute"),
                          accessType : new SFString(field.ACCESSTYPE_INPUTONLY)}),

                        new field({
                          type : field.TYPE_SFFLOAT,
                          name : new SFString("aspectRatio"),
                          accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                          appinfo : new SFString("assumed ratio height/width")}),

                        new field({
                          type : field.TYPE_SFVEC3F,
                          name : new SFString("position_changed"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),

                        new field({
                          type : field.TYPE_SFROTATION,
                          name : new SFString("orientation_changed"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),

                        new field({
                          type : field.TYPE_MFVEC3F,
                          name : new SFString("spine_changed"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),

                        new field({
                          type : field.TYPE_MFVEC2F,
                          name : new SFString("scale_changed"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),

                        new field({
                          type : field.TYPE_MFVEC3F,
                          name : new SFString("point_changed"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),

                        new field({
                          type : field.TYPE_SFBOOL,
                          name : new SFString("trace"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY)}),
                      IS : new SFNode(
                        new IS({
                          connect : new MFNode([
                            new connect({
                              nodeField : new SFString("visible"),
                              protoField : new SFString("visible")}),

                            new connect({
                              nodeField : new SFString("ViewpointNode"),
                              protoField : new SFString("ViewpointNode")}),

                            new connect({
                              nodeField : new SFString("NavigationInfoNode"),
                              protoField : new SFString("NavigationInfoNode")}),

                            new connect({
                              nodeField : new SFString("aspectRatio"),
                              protoField : new SFString("aspectRatio")}),

                            new connect({
                              nodeField : new SFString("trace"),
                              protoField : new SFString("trace")})])})])}),

                    new ROUTE({
                      fromField : new SFString("visibilitySwitchSelection"),
                      fromNode : new SFString("GeometryComputationScript"),
                      toField : new SFString("whichChoice"),
                      toNode : new SFString("VisibilitySwitch")}),

                    new ROUTE({
                      fromField : new SFString("position_changed"),
                      fromNode : new SFString("GeometryComputationScript"),
                      toField : new SFString("translation"),
                      toNode : new SFString("PositionTransform")}),

                    new ROUTE({
                      fromField : new SFString("orientation_changed"),
                      fromNode : new SFString("GeometryComputationScript"),
                      toField : new SFString("rotation"),
                      toNode : new SFString("OrientationTransform")}),

                    new ROUTE({
                      fromField : new SFString("spine_changed"),
                      fromNode : new SFString("GeometryComputationScript"),
                      toField : new SFString("set_spine"),
                      toNode : new SFString("FrustumExtrusion")}),

                    new ROUTE({
                      fromField : new SFString("scale_changed"),
                      fromNode : new SFString("GeometryComputationScript"),
                      toField : new SFString("set_scale"),
                      toNode : new SFString("FrustumExtrusion")}),

                    new ROUTE({
                      fromField : new SFString("point_changed"),
                      fromNode : new SFString("GeometryComputationScript"),
                      toField : new SFString("point"),
                      toNode : new SFString("FrustumCoordinate")})])}))}),
          /*Example use is in separate scene*/

            new Anchor({
              description : new SFString("ViewFrustum Example"),
              url : new MFString(["ViewFrustumExample.x3d"]),
              children : new MFNode([
                new Shape({
                  appearance : new SFNode(
                    new Appearance({
                      material : new SFNode(
                        new Material({
                          diffuseColor : new SFColor([0.8,0.4,0])}))})),
                  geometry : new SFNode(
                    new Text({
                      string : new MFString(["ViewFrustumPrototype.x3d","is a Prototype declaration file.","For an example scene using the prototype,","click this text and view","ViewFrustumExample.x3d"]),
                      fontStyle : new SFNode(
                        new FontStyle({
                          justify : new MFString(["MIDDLE","MIDDLE"]),
                          size : new SFFloat(0.8)}))}))})])})])}))});
console.log(X3D0.toXMLNode());
